module/os/linux/zfs/zpl_inode.c

```diff
@@ -422,9 +430,14 @@ zpl_mkdir(struct inode *dir, struct dentry *dentry, umode_t mode)
 	spl_fstrans_unmark(cookie);
 	kmem_free(vap, sizeof (vattr_t));
 	crfree(cr);
-	ASSERT3S(error, <=, 0);
 
+err:
+	ASSERT3S(error, <=, 0);
+#if defined(HAVE_IOPS_MKDIR_DENTRY)
+	return (ERR_PTR(error));
+#else
 	return (error);
```

Comment:
this is probably going to be more complicated, see current `Documentation/filesystems/vfs.rst` and the end of `Documentation/filesystems/porting.rst`, that `mkdir` note in `vfs.rst` is now dangerously longer and related to NFS, there be dragons, best to be done as they say :)

`d_instantiate_new` should now be used (also in `zpl_create` and probably rename which create whiteouts might need to be checked too), or in less recommended way `unlock_new_inode()` should be after `d_instantiate`

the dentry should be returned if there's a success if I understand it correctly, not sure in how many of the above mentioned case but seems to hold for `mkdir`

Reply:
as for the `unlock_new_inode` thing... I have done that already, https://github.com/openzfs/zfs/pull/16772, can we get that merged please :)

Reply:
Ahh, the error return should probably be `(error ? ERR_PTR(error) : NULL)`, minor thinko.

Changing to `d_instantiate_new()` is _probably_ right, but that's been around since 4.17, so I'm not seeing how its a change for 6.15 as such.

Reply:
Take a look at `git diff v6.14..linus/master Documentation/filesystems/porting.rst`, it talks about this :) I dunno, but from #16788 I would say it's best to follow these two docs (`porting.rst` and `vfs.rst`), because it's a bit of a nightmare to chase it after quite some time goes by and it actually starts to matter in more significant ways - big changes come over the span of multiple releases, it might not matter immediately, but the real reason for the note to be there might come in after a few additional merge windows.

I could also change #16772 to include this `d_instantiate` -> `d_instantiate_new` thing so you don't have to dig `unlock_new_inode` out yourself. Would make sense to me (and I'd also take a look at `zpl_create` and the whiteouts and stuff)

Reply:
IMHO, for the code as it is now, it would make sense to return the dentry if it's a mission success, instead of returning NULL, `d_instantiate_new` can wait

Reply:
I've been neck-deep inode and dentry code for unrelated reasons, so I feel like I get a little more what this is all about :)

So I agree we should switch over to `d_instantiate_new`, and maybe I'll go do that this week. But, as it currently stands, I don't see why we'd change to returning the dentry. It seems to be only necessary if the inode might not be `I_NEW`, but I don't see how it couldn't be, because we always create it.

Reading `vfs_mkdir(), I think we can't just return the same dentry, because if you return one, it will `dput()` the old one. So at least, we'd have to `dget()` before returning, but maybe more.